src/v/storage/disk_log_impl.cc

```diff
@@ -79,9 +79,9 @@ disk_log_impl::~disk_log_impl() {
     vassert(_closed, "log segment must be closed before deleting:{}", *this);
 }
 
-void disk_log_impl::recompute_max_segment_size() {
-    _max_segment_size = internal::jitter_segment_size(
-      std::min(max_segment_size(), segment_size_hard_limit));
+size_t disk_log_impl::compute_max_segment_size() {
+    auto segment_size = std::min(max_segment_size(), segment_size_hard_limit);
+    return segment_size * (1 + _segment_size_jitter);
```

Comment:
reducing the cost of the fast path is great, but it is not clear that removing an invocation of the RNG (integer math, presumably) is better than now having to invoke the FPU. There are also other RNG engines to select from...

Reply:
Thats interesting, I just assumed that the RNG math is complex enough and slower than the floating point multiplication here. I wrote a quick benchmark to validate that now that you mentioned it.. something like as follows..

without 0b84ac6c51ac040168558dcbbef64f88bde703cc :...
```
size_t jitter_segment_size_mt19937(size_t sz, storage::internal::jitter_percents jitter_percents = storage::internal::default_segment_size_jitter) {
    vassert(
      jitter_percents >= 0 || jitter_percents <= 100,
      "jitter percents should be in range [0,100]. Requested {}",
      jitter_percents);
    // multiply by 10 to increase resolution
    auto p = jitter_percents() * 10;
    std::uniform_int_distribution<int64_t> dist(-p, p);
    auto jit = dist(mt19937);
    int64_t jitter = jit * static_cast<int64_t>(sz) / 1000;
    return jitter + sz;
}

PERF_TEST(jitter_test, jitter_test_with_mt19937) {
    constexpr size_t segment_size = 1_GiB;
    perf_tests::start_measuring_time();
    auto x = jitter_segment_size_mt19937(segment_size); 
    perf_tests::do_not_optimize(x);
    perf_tests::stop_measuring_time();
}
```

With 0b84ac6c51ac040168558dcbbef64f88bde703cc:...
```
PERF_TEST(jitter_test, jitter_test_with_float_mul) {
    const auto jitter = storage::internal::random_jitter();
    constexpr size_t segment_size = 1_GiB;
    perf_tests::start_measuring_time();
    auto x = segment_size * (1 + jitter);
    perf_tests::do_not_optimize(x);
    perf_tests::stop_measuring_time();
}
```
Results:..

```
test                                      iterations      median         mad         min         max      allocs       tasks        inst
jitter_test.jitter_test_with_float_mul      19159811    20.659ns     0.031ns    20.563ns    20.769ns       0.000       0.000         0.0
jitter_test.jitter_test_with_lcg            19000981    30.723ns     0.112ns    30.559ns    30.834ns       0.000       0.000         0.0
jitter_test.jitter_test_with_mt19937        17173592    36.254ns     0.074ns    36.180ns    36.541ns       0.000       0.000         0.0
jitter_test.jitter_test_with_mt19937_64     18349613    32.331ns     0.207ns    32.095ns    32.557ns       0.000       0.000         0.0
jitter_test.jitter_test_with_ranlux24        8698590    92.686ns     0.479ns    92.150ns    94.360ns       0.000       0.000         0.0
jitter_test.jitter_test_with_ranlux24_base    17922317    33.477ns     0.037ns    33.437ns    33.530ns       0.000       0.000         0.0
jitter_test.jitter_test_with_ranlux48        3743730   244.807ns     0.447ns   244.293ns   245.648ns       0.000       0.000         0.0
jitter_test.jitter_test_with_ranlux48_base    17991959    33.558ns     0.017ns    33.541ns    33.781ns       0.000       0.000         0.0
jitter_test.jitter_test_with_knuth_b        18474125    32.306ns     0.112ns    32.052ns    32.988ns       0.000       0.000         0.0

```

Reply:
even if the FPU has tiny latency, aren't there a limited number of FPUs so contention is a concern and doesn't the FPU need to trap into the kernel, thereby potentially creating contention and doing things like affecting cache lines?

Reply:
you know of a way to measure this? I can try running some experiments to get some data. 

> contention is a concern

I thought we are not too heavy on floating point arithmetic and the code I added here runs once per a log roll. With a default segment size of 1GiB and 10k partitions oer shard coupled with random jitter, feels like the extra throughput overhead is almost negligible. 

> doesn't the FPU need to trap into the kernel,

Curious how's this different from a typical ALU. I'm not too familiar with inner workings of an FPU.

Reply:
> I thought we are not too heavy on floating point arithmetic and the code I added here runs once per a log roll.

It caught my eye because it didn't seem to be consistent with the commit message "Avoids repeated RNG calls in the hot path."

In any case, I don't think there is any need to measure and change this now, but it is something to keep in mind. We can ask Travis D for some pointers to information. I suspect you can find similar information if you look up the reason why floating point arithmetic isn't used in the kernel.